#2051 Internationalize Distances Dates Times

MattJ12 opened the pull request:
Resolves #2051 

This PR internationalizes the distances, dates, and times on Project Sidewalk using the built in language detection in the Play Framework.  the [i18next](https://www.i18next.com/) and [Moment](https://momentjs.com/docs/) libraries are used to handle some additional translations, although all language detection is based off of that in the Play Framework.  Below is a list with pictures of the main areas where translations were implemented to reflect the user's country:

- User mission distance on the audit page (metric/IS) depending on what system the user's country uses
Before:
![Mission_Start_old](https://user-images.githubusercontent.com/32688680/83315797-36580480-a1d7-11ea-922e-423556f846d3.png)
![explore_screen_old](https://user-images.githubusercontent.com/32688680/83315805-440d8a00-a1d7-11ea-9347-0f351ce260ee.png)
![mission_complete_old](https://user-images.githubusercontent.com/32688680/83315811-4a9c0180-a1d7-11ea-889e-3c7524f5ec5b.png)

After:
![mission_start_new](https://user-images.githubusercontent.com/32688680/83315815-52f43c80-a1d7-11ea-93ef-783b03e6060e.png)
![current_mission_new](https://user-images.githubusercontent.com/32688680/83315826-5c7da480-a1d7-11ea-9a61-e51cc1e69770.png)
![mission_complete_new](https://user-images.githubusercontent.com/32688680/83315835-630c1c00-a1d7-11ea-9163-89901456cdc3.png)

- Total distance audited animation on the landing page

Before:
![landing_page_correct_new](https://user-images.githubusercontent.com/32688680/83316183-54266900-a1d9-11ea-8da9-fe0fbd529709.png)

After:
![audited_distance_landing_page](https://user-images.githubusercontent.com/32688680/83315903-c4cc8600-a1d7-11ea-88ac-01ac13f91110.png)

- Distances shown on the choropleth pop-ups
Before:
![Choropleth_pop_up_old](https://user-images.githubusercontent.com/32688680/83315399-9305f000-a1d4-11ea-9c40-9d416883a5dd.png)

After:
![Choropleth_pop_up](https://user-images.githubusercontent.com/32688680/83315367-576b2600-a1d4-11ea-860c-fac5a044c7c9.png)

- Format and words in the footer and mini-footer.
Before:
![footer_old](https://user-images.githubusercontent.com/32688680/83315935-01987d00-a1d8-11ea-891a-77e05a00b941.png)

After:
![footer_new](https://user-images.githubusercontent.com/32688680/83315941-0826f480-a1d8-11ea-845b-d8ad4c1b838b.png)

- Format and words in the label-box pop-up on the How To Label page
Before:
The dates were not translated and in the wrong format for the specific country

After:
![label_map_new](https://user-images.githubusercontent.com/32688680/83316056-9ac79380-a1d8-11ea-9bf2-3110b363905a.png)

misaugstad commented:
Great!! Very exciting stuff!! Can you just add some before/after images to the PR description? :)

MattJ12 commented:
> Great!! Very exciting stuff!! Can you just add some before/after images to the PR description? :)

Thanks!  I added some before after pictures above for each area that I internationalized. 

I also noticed a couple possible issues that would be outside the scope of this PR:

- On the mission complete modal the distance types (kilometers/miles) in the bottom right are always plural, which would not be correct if there was one unit.

- On the label map pop-up I translated the date/time to have the correct words and be in the correct format for the user's country.  However, the other information in that pop-up is not translated.

misaugstad commented:
So I tested this PR briefly just now. I went through most of the pages in en-US, en, and es. Here are the issues I came across:

Landing page
1. At the beginning of the "your work is making a difference" section, it still says "X miles" in en and "X millas" in es.
1. Below that, I'm seeing "0.3 miles covered" in en-US, "0.3 kilometers covered" in en, and "0.5 kilómetros cubiertos" in es. So the number is not being converted from miles to km in en.
1. In the footer, the date format is currently the same for en and en-US. I'm assuming that the format we used is not standard outside the US. Looks good in Spanish though.

User dashboard:
1. On the choropleth popups, I don't believe that miles are being converted to kilometers. The name of the unit is changing, but I don't think the conversion is happening.
1. You had already pointed out the dates on the graph not being in the right format.
1. In the table, the date format is the same for en and en-US, which is probably not correct. And when it is in Spanish, the length of the string seems to be too long and it is messing up the formatting of the table. Can you either switch it to a shorter format or make whatever changes are needed in the table so that it doesn't mess up the formatting? I'll add a screenshot at the bottom of this comment.

Lable map:
1. The dates/times are the same for en and en-US, which is probably not right.

Restuls:
Looks good!

Audit:
1. In the little "current neighborhood" section in the top right, it looks like when going to en it converts from miles to km, but it keeps the unit as saying miles.
1. On the mission complete modal, it looks like some (if not all) of the distances are not being converted. The units are changing, they just aren't being converted.

![messed-up-user-dashboard-table](https://user-images.githubusercontent.com/6518824/83567472-791b2480-a4d6-11ea-9914-d67b7de9296b.png)

misaugstad commented:
Oh right, and on the admin page, I'm seeing some errors in the console about moment.js. I didn't really test the admin page, just wanted to see what would happen and saw some errors in console.

MattJ12 commented:
> So I tested this PR briefly just now. I went through most of the pages in en-US, en, and es. Here are the issues I came across:
> 
> Landing page
> 
> 1. At the beginning of the "your work is making a difference" section, it still says "X miles" in en and "X millas" in es.
> 2. Below that, I'm seeing "0.3 miles covered" in en-US, "0.3 kilometers covered" in en, and "0.5 kilómetros cubiertos" in es. So the number is not being converted from miles to km in en.
> 3. In the footer, the date format is currently the same for en and en-US. I'm assuming that the format we used is not standard outside the US. Looks good in Spanish though.
> 
> User dashboard:
> 
> 1. On the choropleth popups, I don't believe that miles are being converted to kilometers. The name of the unit is changing, but I don't think the conversion is happening.
> 2. You had already pointed out the dates on the graph not being in the right format.
> 3. In the table, the date format is the same for en and en-US, which is probably not correct. And when it is in Spanish, the length of the string seems to be too long and it is messing up the formatting of the table. Can you either switch it to a shorter format or make whatever changes are needed in the table so that it doesn't mess up the formatting? I'll add a screenshot at the bottom of this comment.
> 
> Lable map:
> 
> 1. The dates/times are the same for en and en-US, which is probably not right.
> 
> Restuls:
> Looks good!
> 
> Audit:
> 
> 1. In the little "current neighborhood" section in the top right, it looks like when going to en it converts from miles to km, but it keeps the unit as saying miles.
> 2. On the mission complete modal, it looks like some (if not all) of the distances are not being converted. The units are changing, they just aren't being converted.
> 
> ![messed-up-user-dashboard-table](https://user-images.githubusercontent.com/6518824/83567472-791b2480-a4d6-11ea-9914-d67b7de9296b.png)

I addressed most of the issues here.  On the choropleth, I edited the code in the same way as the accessibility choropleth on the results page.  Also, Moment.js uses en-us translations for en translations, so they should use the same format when translating dates.  In the future we would just define more translations as we need them such as en-au.

misaugstad commented:
>Moment.js uses en-us translations for en translations, so they should use the same format when translating dates. In the future we would just define more translations as we need them such as en-au.

That makes sense.

Okay I did another round of testing. I still found a few obvious errors. Please try to be a bit more thorough in your testing in the future, it will save both of us a lot of time!

Errors:
1. Landing page, your work is making a difference section. The distance is still not being converted in either location.
1. Landing page, your work is making a difference section. In one of the locations, a "," is being used instead of "." for decimal numbers in Spanish. But just in one location so it is not consistent. I think that in Spain they use "," but in Mexico they typically use "." like the US. I think to limit the scope of the PR, you could just use ".", and someone in the future can internationalize that piece.
1. User dashboard choropleth: distance is not being converted on popup.
1. On the validation page we have the month and year that the image was taken in the bottom left corner of the pano. This is not being changed for different locales.
1. Admin page: still getting errors in the console about Moment.js. Are you not getting these?
1. Admin page, analytics tab: on the choropleth, I clicked on a neighborhood and it said "298% complete". It is supposed to be 3%, so it looks like it's getting multiplied by 100.
1. Admin page, analytics tab: there are a few graphs that have dates on the x-axis, same as on the user dashboard. Would you be able to have these dates change for different locales? I peaked at the user dashboard code for it and it looks like it was pretty simple. If it isn't easy to do and you're running out of time, feel free to skip this bit b/c the admin page is less important.

Other things I've realized we want:
1. For the user dashboard (and admin dashboard) we are using dates in the format MM/DD/YYYY, and the Spanish version is using DD/MM/YYYY. Is there a less ambiguous date format that is still really short? I think it would be best if we could avoid the MM/DD/YYYY format to prevent ambiguity.  If you're running short on time, you can skip the implementation of this (and let me know that's what you're doing so I can create an issue for it).
1. On the audit page, can you change all the places where the number and the unit do not have a space between them to having a space? So instead of "1000ft" we would have "1000 ft". It happens in at least 3 places on the audit page.

MattJ12 commented:
> > Moment.js uses en-us translations for en translations, so they should use the same format when translating dates. In the future we would just define more translations as we need them such as en-au.
> 
> That makes sense.
> 
> Okay I did another round of testing. I still found a few obvious errors. Please try to be a bit more thorough in your testing in the future, it will save both of us a lot of time!
> 
> Errors:
> 
> 1. Landing page, your work is making a difference section. The distance is still not being converted in either location.
> 2. Landing page, your work is making a difference section. In one of the locations, a "," is being used instead of "." for decimal numbers in Spanish. But just in one location so it is not consistent. I think that in Spain they use "," but in Mexico they typically use "." like the US. I think to limit the scope of the PR, you could just use ".", and someone in the future can internationalize that piece.
> 3. User dashboard choropleth: distance is not being converted on popup.
> 4. On the validation page we have the month and year that the image was taken in the bottom left corner of the pano. This is not being changed for different locales.
> 5. Admin page: still getting errors in the console about Moment.js. Are you not getting these?
> 6. Admin page, analytics tab: on the choropleth, I clicked on a neighborhood and it said "298% complete". It is supposed to be 3%, so it looks like it's getting multiplied by 100.
> 7. Admin page, analytics tab: there are a few graphs that have dates on the x-axis, same as on the user dashboard. Would you be able to have these dates change for different locales? I peaked at the user dashboard code for it and it looks like it was pretty simple. If it isn't easy to do and you're running out of time, feel free to skip this bit b/c the admin page is less important.
> 
> Other things I've realized we want:
> 
> 1. For the user dashboard (and admin dashboard) we are using dates in the format MM/DD/YYYY, and the Spanish version is using DD/MM/YYYY. Is there a less ambiguous date format that is still really short? I think it would be best if we could avoid the MM/DD/YYYY format to prevent ambiguity.  If you're running short on time, you can skip the implementation of this (and let me know that's what you're doing so I can create an issue for it).
> 2. On the audit page, can you change all the places where the number and the unit do not have a space between them to having a space? So instead of "1000ft" we would have "1000 ft". It happens in at least 3 places on the audit page.

I addressed the requested changes except error number 7, which we decided to leave for another PR.  I added support for translating the date on the bottom left of the GSV Pano for auditing, not just validating.

misaugstad commented:
Great, thanks @MattJ12 ! This round of testing went really well!! Only the tiniest little notes:
1. On the landing page in the "your work is making a difference" section, if the number of miles/km audited goes into the thousands, there is only a comma in the 2nd time the distance is listed. So mine says "...already mapped 1012 kilometers of..." but then "1,011.8 kilometers covered". Could you make it so that there is a comma in both places for consistency?
1. For the label popups on /labelMap and the like, the image date for the pano should only include month and year instead of day as well. Only the month is recorded, so when you include the day it is just getting set to the last day of the month.
1. There's a couple places where you use the full month name that I'm hoping we can switch to abbreviated month names if that's not too hard? The three places are the bottom left corner of both the validation and audit panos, and then on the chart on the user dashboard.

I'll make sure to get to reading the code tomorrow!

MattJ12 commented:
> Great, thanks @MattJ12 ! This round of testing went really well!! Only the tiniest little notes:
> 
> 1. On the landing page in the "your work is making a difference" section, if the number of miles/km audited goes into the thousands, there is only a comma in the 2nd time the distance is listed. So mine says "...already mapped 1012 kilometers of..." but then "1,011.8 kilometers covered". Could you make it so that there is a comma in both places for consistency?
> 2. For the label popups on /labelMap and the like, the image date for the pano should only include month and year instead of day as well. Only the month is recorded, so when you include the day it is just getting set to the last day of the month.
> 3. There's a couple places where you use the full month name that I'm hoping we can switch to abbreviated month names if that's not too hard? The three places are the bottom left corner of both the validation and audit panos, and then on the chart on the user dashboard.
> 
> I'll make sure to get to reading the code tomorrow!

Done!  Let me know if you like the way the abbreviations look.

misaugstad commented:
Testing wise, just noticed that the date in the bottom left corner of audit pano is always the current month and year. I think the issue is that you accidentally wrote `moment(data.imageData)` instead of `moment(data.imageDate)`.

Also getting errors in the console on the user dashboard and validation pages because there's no validation.json or dashboard.json for en-US. Can you add those json files so we don't have that error anymore? They probably just need an open and closing bracket in them? Just test to make sure there is no error there anymore :)

>Done! Let me know if you like the way the abbreviations look.

Abbreviations look good, thank you!

Okay going to begin diving into the code now.

misaugstad commented:
Actually, before I review, could you do one thorough pass over all your code while looking at the [style guide](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Style-Guide)?

MattJ12 commented:
@misaugstad I fixed the issue with the audit GSV Panorama not updating the date as well as the missing namespaces for en-US.  I also went through and fixed the style issues I could find.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/navigation/MapService.js:
```diff
@@ -864,13 +860,13 @@ function MapService (canvas, neighborhoodModel, uiMap, params) {
     }
 
-    function finishCurrentTaskBeforeJumping(mission, nextTask) {
+    function finishCurrentTaskBeforeJumping(mission) {
```
I think these are accidental reverts from develop too

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/status/StatusFieldMission.js:
```diff
@@ -40,22 +49,31 @@ StatusFieldMission.prototype._distanceToString = function (distance, unit) {
+        return "750 ft";
     } else if (distance === "0.1894") {
-        return "1000ft";
+        if (distanceType === "metric") return this.convertToMetric(1000);
```
Same comment about removing distance after 1,000 feet.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/task/TaskContainer.js:
```diff
@@ -76,7 +76,7 @@ function TaskContainer (navigationModel, neighborhoodModel, streetViewService, s
      * End the current task.
      */
-    self.endTask = function (task, nextTask) {
+    self.endTask = function (task) {
```
More stuff from develop

misaugstad commented on public/javascripts/SVValidate/css/svv-menu-buttons.css:
```diff
@@ -12,22 +12,6 @@
     justify-content: space-between;
 }
 
-.validation-comment-box {
```
from develop

MattJ12 commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionComplete.js:
```diff
@@ -269,10 +269,13 @@ function ModalMissionComplete (svl, missionContainer, missionModel, taskContaine
         // Update the horizontal bar chart to show how much distance the user has audited
-        var unit = {units: 'miles'};
+        var measurementSystem = i18next.t('measurement-system');
+        var unit = {units: "miles"};
```
I used dot access to get a String for the `mission.getDistance` method.  I didn't convert the variable unit to a String because in later lines there are method such as `completedLineDistance` and `completedLineDistanceAcrossAllUsersUsingPriority` that I think have to take an object as parameter.

misaugstad replied:
Ahh sorry, read this stuff incorrectly, you're right.

MattJ12 commented:
@misaugstad I addressed the comments.  The major areas of update were:

- How distances are converted on choropleths

- Mission distance ranges and unit abbreviations for audit

- Moving more distance types to the `common.json` namespace file to make locale processing easier in the js files

misaugstad commented on app/controllers/ApplicationController.scala:
```diff
@@ -132,7 +134,11 @@ class ApplicationController @Inject() (implicit val env: Environment[User, Sessi
                 (otherName + ", " + otherState, otherURL)
               }
-              Future.successful(Ok(views.html.index("Project Sidewalk", Some(user), cityName, stateAbbreviation, cityShortName, mapathonLink, cityStr, otherCityUrls)))
+              // Get total audited distance. If using metric system, convert from miles to kilometers
```
Comment needs a period at the end.

misaugstad commented on app/controllers/ApplicationController.scala:
```diff
@@ -132,7 +134,11 @@ class ApplicationController @Inject() (implicit val env: Environment[User, Sessi
               }
-              Future.successful(Ok(views.html.index("Project Sidewalk", Some(user), cityName, stateAbbreviation, cityShortName, mapathonLink, cityStr, otherCityUrls)))
+              // Get total audited distance. If using metric system, convert from miles to kilometers
+              val auditedDistance =
```
Can you add a type declaration here? I.e., `val auditedDistance: Float =`

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMission.js:
```diff
@@ -101,6 +101,15 @@ function ModalMission (missionContainer, neighborhoodContainer, uiModalMission,
     };
 
+    /**
+     *  This method takes in an integer feet and converts it to meters, truncuating all decimals.
```
truncuating -> truncating

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMission.js:
```diff
@@ -218,23 +227,21 @@ ModalMission.prototype._distanceToString = function  (distance, unit) {
-        return "500ft";
+    if (distance === "0.0947"){
+        if (distanceType === "metric") return this.convertToMetric(500, unitAbbreviation);
+        return "500 " + unitAbbreviation;
```
Can you update all these to use `else return ...` instead of just `return ...`

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionComplete.js:
```diff
@@ -268,11 +268,12 @@ function ModalMissionComplete (svl, missionContainer, missionModel, taskContaine
     this.update = function (mission, neighborhood) {
-        // Update the horizontal bar chart to show how much distance the user has audited
-        var unit = {units: 'miles'};
+        // Update the horizontal bar chart to show how much distance the user has audited.
```
"how much distance the user has audited" -> "how much the user has audited" or "the distance the user has audited"

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/status/StatusFieldMission.js:
```diff
@@ -29,6 +29,15 @@ function StatusFieldMission (modalModel, uiStatusField) {
     };
+
+    /**
+     *  This method takes in an integer feet and converts it to meters, truncuating all decimals.
```
truncuating -> truncating

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/status/StatusFieldMission.js:
```diff
@@ -40,22 +49,20 @@ StatusFieldMission.prototype._distanceToString = function (distance, unit) {
     if (distance === "0.0947"){
-        return "500ft";
+        if (distanceType === "metric") return this.convertToMetric(500, unitAbbreviation);
+        return "500 " + unitAbbreviation;
```
Same with this section and adding the `else`

misaugstad commented on public/javascripts/TimestampLocalization.js:
```diff
@@ -1,27 +1,26 @@
-        if (localDate.format(format) !== "Invalid date") {
-            this.textContent = localDate.format(format);
-        }
+function updateTimestamp(locale) {
```
Can you add a short comment saying what this function is for? (converting to local time and putting into locale's format).

misaugstad replied:
Can you also change the name to be `updateTimestamps`

misaugstad commented on public/javascripts/TimestampLocalization.js:
```diff
@@ -1,27 +1,26 @@
+            // Load correct locale.
+            moment.locale(locale);
+    
+            let localDate = moment(new Date(timestampText + " UTC"));
```
Does this just convert to local time? Or does it also change the formatting? If it does both, could you have a comment here? Maybe just have it look like:
```
            // Converts to local time and changes to local time/date format.
            moment.locale(locale);
            let localDate = moment(new Date(timestampText + " UTC"));
```

MattJ12 commented:
> This looks SO much better. We are basically there. I've added just the most minor of comments to address. We should be merging once you've addressed these.

Alright, got them fixed!
